resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java

```diff
@@ -268,6 +270,16 @@ static void waitForPermission(final RateLimiter rateLimiter) throws IllegalState
      */
     boolean getPermission(Duration timeoutDuration);
 
+    /**
+     * Reserves a permission from this rate limiter and returns nanoseconds you should wait for it.
+     * If returned long is negative, it means that you failed to reserve permission,
+     * possibly your {@code timeoutDuration} is less then time to wait for permission.
+     *
+     * @param timeoutDuration the maximum time you want to wait.
+     * @return {@code long} amount of nanoseconds you should wait for reserved permission. if negative, it means you failed to reserve.
+     */
+    long reservePermission(Duration timeoutDuration);
```

Comment:
From an API-perspective I would prefer to return a Duration instead of a long. But I guess this would lead to more. GC

Reply:
Yep, it's a low level API where we want to avoid unnecessary GC pressure.